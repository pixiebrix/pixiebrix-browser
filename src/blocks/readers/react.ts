/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
