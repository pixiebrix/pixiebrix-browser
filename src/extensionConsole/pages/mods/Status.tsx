/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Status.module.scss";

import React from "react";
import { Button } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCheck,
  faExclamationTriangle,
  faPause,
  faShieldAlt,
  faSync,
} from "@fortawesome/free-solid-svg-icons";
import AsyncButton from "@/components/AsyncButton";
import { type ModViewItem } from "@/types/modTypes";
import useModPermissions from "@/mods/hooks/useModPermissions";
import reportEvent from "@/telemetry/reportEvent";
import { Events } from "@/telemetry/events";
import { useHistory } from "react-router";
import useActivatedModComponents from "@/mods/hooks/useActivatedModComponents";
import { API_PATHS } from "@/data/service/urlPaths";

const Status: React.VoidFunctionComponent<{
  modViewItem: ModViewItem;
}> = ({ modViewItem }) => {
  const history = useHistory();

  const {
    modId,
    hasUpdate,
    status,
    activatedModVersion,
    isUnavailable,
    modActions: { showActivate, showReactivate },
  } = modViewItem;

  const modComponents = useActivatedModComponents(modId);

  const { hasPermissions, requestPermissions } =
    useModPermissions(modComponents);

  if (isUnavailable) {
    return (
      <div className="text-warning">
        <div className={styles.root}>
          <FontAwesomeIcon icon={faExclamationTriangle} />
          <span className={styles.textStatus}>
            Active
            <span className={styles.versionNumber}>No longer available</span>
          </span>
        </div>
      </div>
    );
  }

  if (showActivate) {
    return (
      <Button
        size="sm"
        variant="outline-primary"
        onClick={() => {
          reportEvent(Events.START_MOD_ACTIVATE, {
            modId,
            screen: "extensionConsole",
            reinstall: false,
          });
          history.push(API_PATHS.MOD_ACTIVATE(modId));
        }}
      >
        Activate
      </Button>
    );
  }

  if (hasUpdate && showReactivate) {
    return (
      <Button
        size="sm"
        variant="info"
        onClick={() => {
          reportEvent(Events.START_MOD_ACTIVATE, {
            modId,
            screen: "extensionConsole",
            reinstall: true,
          });
          history.push(API_PATHS.MOD_ACTIVATE(modId, true));
        }}
      >
        <FontAwesomeIcon icon={faSync} /> Update
      </Button>
    );
  }

  if (!hasPermissions) {
    // Use "Allow" for caption because the original "Grant Permissions" was too long
    return (
      <AsyncButton size="sm" variant="info" onClick={requestPermissions}>
        <FontAwesomeIcon icon={faShieldAlt} /> Allow
      </AsyncButton>
    );
  }

  if (status === "Paused") {
    return (
      <div className="text-muted">
        <div className={styles.root}>
          <FontAwesomeIcon icon={faPause} />
          <span className={styles.textStatus}>
            Paused
            <span className={styles.versionNumber}>
              version {activatedModVersion}
            </span>
          </span>
        </div>
      </div>
    );
  }

  return (
    <div className="text-success">
      <div className={styles.root}>
        <FontAwesomeIcon icon={faCheck} />
        <span className={styles.textStatus}>
          Active
          <span className={styles.versionNumber}>
            version {activatedModVersion}
          </span>
        </span>
      </div>
    </div>
  );
};

export default Status;
