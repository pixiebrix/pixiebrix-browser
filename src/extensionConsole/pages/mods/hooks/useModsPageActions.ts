/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import useActivateAction from "@/extensionConsole/pages/mods/hooks/useActivateAction";
import useViewPublishAction from "@/extensionConsole/pages/mods/hooks/useViewPublishAction";
import useMarketplaceUrl from "@/mods/hooks/useMarketplaceUrl";
import useViewShareAction from "@/extensionConsole/pages/mods/hooks/useViewShareAction";
import useDeleteExtensionAction from "@/mods/hooks/useDeleteExtensionAction";
import useReactivateAction from "@/extensionConsole/pages/mods/hooks/useReactivateAction";
import { type ModViewItem } from "@/types/modTypes";
import useRequestPermissionsAction from "@/mods/hooks/useRequestPermissionsAction";
import useViewLogsAction from "@/extensionConsole/pages/mods/hooks/useViewLogsAction";
import useDeactivateAction from "@/mods/hooks/useDeactivateAction";

type ActionCallback = () => void;

export type ModsPageActions = {
  reactivate: ActionCallback | null;
  activate: ActionCallback | null;
  viewPublish: ActionCallback | null;
  viewInMarketplaceHref: string | null;
  viewShare: ActionCallback | null;
  deleteExtension: ActionCallback | null;
  deactivate: ActionCallback | null;
  viewLogs: ActionCallback | null;
  requestPermissions: ActionCallback | null;
};

function useModsPageActions(modViewItem: ModViewItem): ModsPageActions {
  const marketplaceListingUrl = useMarketplaceUrl(modViewItem);
  const viewPublish = useViewPublishAction(modViewItem);
  const viewShare = useViewShareAction(modViewItem);
  const reactivate = useReactivateAction(modViewItem);
  const viewLogs = useViewLogsAction(modViewItem);
  const activate = useActivateAction(modViewItem);
  const deactivate = useDeactivateAction(modViewItem);
  const deleteExtension = useDeleteExtensionAction(modViewItem);
  const requestPermissions = useRequestPermissionsAction(modViewItem);

  return {
    viewPublish,
    viewInMarketplaceHref: marketplaceListingUrl,
    viewShare,
    deleteExtension,
    deactivate,
    reactivate,
    viewLogs,
    activate,
    requestPermissions,
  };
}

export default useModsPageActions;
