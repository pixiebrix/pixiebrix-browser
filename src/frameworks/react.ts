/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import fromPairs from "lodash/fromPairs";

interface DOMFiber {
  _instance: any;
  _currentElement: {
    _owner: DOMFiber;
  };
}

interface ComponentFiber {
  return: ComponentFiber;
  memoizedProps: { [key: string]: unknown };
}

export class ComponentNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ComponentNotFoundError";
  }
}

export function readReactProps(
  fiber: ComponentFiber,
  rootProp?: string
): Record<string, unknown> {
  // getComponentFiber is already returning the "return" for the component
  const memoizedProps = fiber.memoizedProps;
  console.debug("Memoized props", memoizedProps);
  const props = rootProp ? memoizedProps[rootProp] : memoizedProps;
  return fromPairs(Object.entries(props).filter(([key]) => key !== "children"));
}

export function findReactComponent(
  dom: HTMLElement,
  traverseUp = 0
): ComponentFiber {
  // https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging
  // https://stackoverflow.com/a/39165137/402560

  const key = Object.keys(dom).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );
  // @ts-ignore: works in practice
  const domFiber: DOMFiber | null = dom[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found at selector");
  }

  // react <16
  // TODO: find a react 15.0 app to test this on
  if (domFiber._currentElement) {
    throw new Error("Support for React <16 not implemented");
    // let componentFiber = domFiber._currentElement._owner;
    // for (let i = 0; i < traverseUp; i++) {
    //     componentFiber = componentFiber._currentElement._owner;
    // }
    // return componentFiber._instance;
  }

  // react 16+
  const getComponentFiber = (fiber: any) => {
    //return fiber._debugOwner; // this also works, but is __DEV__ only
    let parentFiber = fiber.return;
    while (typeof parentFiber.type == "string") {
      parentFiber = parentFiber.return;
    }
    return parentFiber;
  };
  let componentFiber = getComponentFiber(domFiber);
  for (let i = 0; i < traverseUp; i++) {
    componentFiber = getComponentFiber(componentFiber);
  }
  return componentFiber;
}
