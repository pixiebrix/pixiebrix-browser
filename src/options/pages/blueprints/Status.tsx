/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Status.module.scss";

import React from "react";
import { Button } from "react-bootstrap";
import { type InstallableViewItem } from "./blueprintsTypes";
import useInstallableViewItemActions from "./useInstallableViewItemActions";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCheck,
  faExclamationTriangle,
  faPause,
  faShieldAlt,
  faSync,
} from "@fortawesome/free-solid-svg-icons";
import AsyncButton from "@/components/AsyncButton";

const Status: React.VoidFunctionComponent<{
  installableViewItem: InstallableViewItem;
}> = ({ installableViewItem }) => {
  const { activate, reinstall, requestPermissions } =
    useInstallableViewItemActions(installableViewItem);

  const { hasUpdate, status, installedVersionNumber, unavailable } =
    installableViewItem;

  if (unavailable) {
    return (
      <div className="text-warning">
        <div className={styles.root}>
          <FontAwesomeIcon icon={faExclamationTriangle} />
          <span className={styles.textStatus}>
            Active
            <span className={styles.versionNumber}>No longer available</span>
          </span>
        </div>
      </div>
    );
  }

  if (activate) {
    return (
      <Button size="sm" variant="outline-primary" onClick={activate}>
        Activate
      </Button>
    );
  }

  if (hasUpdate && reinstall) {
    return (
      <Button size="sm" variant="info" onClick={reinstall}>
        <FontAwesomeIcon icon={faSync} /> Update
      </Button>
    );
  }

  if (requestPermissions) {
    // Use "Allow" for caption because the original "Grant Permissions" was too long
    return (
      <AsyncButton size="sm" variant="info" onClick={requestPermissions}>
        <FontAwesomeIcon icon={faShieldAlt} /> Allow
      </AsyncButton>
    );
  }

  if (status === "Paused") {
    return (
      <div className="text-muted">
        <div className={styles.root}>
          <FontAwesomeIcon icon={faPause} />
          <span className={styles.textStatus}>
            Paused
            {installedVersionNumber && (
              <span className={styles.versionNumber}>
                version {installedVersionNumber}
              </span>
            )}
          </span>
        </div>
      </div>
    );
  }

  return (
    <div className="text-success">
      <div className={styles.root}>
        <FontAwesomeIcon icon={faCheck} />
        <span className={styles.textStatus}>
          Active
          {installedVersionNumber && (
            <span className={styles.versionNumber}>
              version {installedVersionNumber}
            </span>
          )}
        </span>
      </div>
    </div>
  );
};

export default Status;
