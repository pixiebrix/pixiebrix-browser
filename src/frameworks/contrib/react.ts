/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { pickBy } from "lodash";
import { ComponentNotFoundError, ignoreNotFound } from "@/frameworks/errors";
import { RootInstanceVisitor } from "@/frameworks/scanner";
import { ReadableComponentAdapter, traverse } from "@/frameworks/component";
import { isNode } from "@/frameworks/dom";

// React architecture references:
// https://github.com/acdlite/react-fiber-architecture
// https://indepth.dev/posts/1008/inside-fiber-in-depth-overview-of-the-new-reconciliation-algorithm-in-react#memoizedprops
// https://github.com/Venryx/mobx-devtools-advanced/blob/master/Docs/TreeTraversal.md
// https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging

interface RootInstance {
  _reactRootContainer: unknown;
}

/**
 * Structure from:
 * - https://github.com/facebook/react/blob/50d9451f320a9aaf94304209193562cc385567d8/packages/react-devtools-shared/src/backend/legacy/renderer.js#L739
 * - https://github.com/facebook/react/blob/v15.0.0/src/renderers/shared/reconciler/ReactReconciler.js
 */
interface LegacyInstance {
  /**
   * Public instance
   */
  _instance: unknown;

  _currentElement: {
    _owner: LegacyInstance;
  };
}

/**
 * Type definition from: https://github.com/facebook/react/blob/4c6470cb3b821f3664955290cd4c4c7ac0de733a/packages/react-reconciler/src/ReactInternalTypes.js#L62
 */
interface Fiber {
  /**
   * The Fiber to return to after finishing processing this one.
   * This is effectively the parent, but there can be multiple parents (two)
   * so this is only the parent of the thing we're currently processing.
   * It is conceptually the same as the return address of a stack frame.
   */
  return: Fiber | null;

  /**
   * The props used to create the output.
   */
  memoizedProps: { [prop: string]: unknown };

  /**
   * The resolved function/class associated with this fiber.
   */
  type: string | object;

  /**
   * The local state associated with this fiber.
   */
  stateNode: Node | object;
}

export function isManaged(node: Node): boolean {
  return Object.keys(node).some((key) =>
    key.startsWith("__reactInternalInstance$")
  );
}

export function hasReactProps(fiber: Fiber): boolean {
  return Object.keys(fiber.memoizedProps).some((x) => x !== "children");
}

export function readReactProps(fiber: Fiber): { [prop: string]: unknown } {
  return pickBy(fiber.memoizedProps, (value, key) => key !== "children");
}

function getComponentFiber(fiber: Fiber): Fiber {
  // return fiber._debugOwner; // this also works, but is __DEV__ only
  let parentFiber = fiber.return;
  while (typeof parentFiber.type === "string") {
    // string for HTML nodes, so traverse
    parentFiber = parentFiber.return;
  }
  return parentFiber;
}

export function findReactComponent(node: Node, traverseUp = 0): Fiber {
  // https://stackoverflow.com/a/39165137/402560
  const key = Object.keys(node).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );

  const domFiber: Fiber | LegacyInstance | null = (node as any)[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found for element");
  }

  if ("_currentElement" in domFiber) {
    // FIXME: test this for React <16
    const owner = (x: LegacyInstance) => x._currentElement._owner;
    const fiber = traverse(owner, owner(domFiber), traverseUp);
    return fiber._instance as Fiber;
  } else {
    return traverse(getComponentFiber, getComponentFiber(domFiber), traverseUp);
  }
}

export class ReactRootVisitor implements RootInstanceVisitor<RootInstance> {
  public rootInstances: RootInstance[] = [];
  visit(node: Element | Node): boolean {
    if ("_reactRootContainer" in node) {
      this.rootInstances.push(node);
      return false;
    }
    return true;
  }
}

export const adapter: ReadableComponentAdapter<Fiber> = {
  isManaged,
  getComponent: (node) => ignoreNotFound(() => findReactComponent(node, 0)),
  getParent: getComponentFiber,
  getNode: (instance) =>
    isNode(instance.stateNode) ? instance.stateNode : null,
  getData: readReactProps,
  hasData: hasReactProps,
};

export default adapter;
