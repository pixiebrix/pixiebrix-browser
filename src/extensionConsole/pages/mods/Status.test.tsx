/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import Status from "@/extensionConsole/pages/mods/Status";
import useModsPageActions, {
  type ModsPageActions,
} from "@/extensionConsole/pages/mods/hooks/useModsPageActions";

jest.mock("@/extensionConsole/pages/mods/hooks/useModsPageActions", () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({}),
}));

const useModPageActionsMock = useModsPageActions as jest.MockedFunction<
  typeof useModsPageActions
>;

describe("Status", () => {
  beforeEach(() => {
    useModPageActionsMock.mockReturnValue({} as ModsPageActions);
  });

  it("shows active", async () => {
    const { asFragment } = render(
      <Status
        modViewItem={
          {
            active: true,
          } as any
        }
      />
    );

    expect(screen.getByText("Active")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows activate", async () => {
    useModPageActionsMock.mockReturnValue({
      activate: jest.fn(),
    } as unknown as ModsPageActions);

    const { asFragment } = render(
      <Status
        modViewItem={
          {
            active: true,
          } as any
        }
      />
    );

    expect(screen.getByText("Activate")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows warning for unavailable", async () => {
    const { asFragment } = render(
      <Status
        modViewItem={
          {
            unavailable: true,
          } as any
        }
      />
    );

    expect(screen.getByText("No longer available")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("paused", async () => {
    const { asFragment } = render(
      <Status
        modViewItem={
          {
            status: "Paused",
          } as any
        }
      />
    );

    expect(screen.getByText("Paused")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
