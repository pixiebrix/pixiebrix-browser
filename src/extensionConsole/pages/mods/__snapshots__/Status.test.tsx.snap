// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status paused 1`] = `
<DocumentFragment>
  <div
    class="text-muted"
  >
    <div
      class="root"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-pause fa-w-14 "
        data-icon="pause"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M144 479H48c-26.5 0-48-21.5-48-48V79c0-26.5 21.5-48 48-48h96c26.5 0 48 21.5 48 48v352c0 26.5-21.5 48-48 48zm304-48V79c0-26.5-21.5-48-48-48h-96c-26.5 0-48 21.5-48 48v352c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48z"
          fill="currentColor"
        />
      </svg>
      <span
        class="textStatus"
      >
        Paused
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Status shows activate 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline-primary btn-sm"
    type="button"
  >
    Activate
  </button>
</DocumentFragment>
`;

exports[`Status shows active 1`] = `
<DocumentFragment>
  <div
    class="text-success"
  >
    <div
      class="root"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check fa-w-16 "
        data-icon="check"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
      <span
        class="textStatus"
      >
        Active
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Status shows warning for unavailable 1`] = `
<DocumentFragment>
  <div
    class="text-warning"
  >
    <div
      class="root"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-exclamation-triangle fa-w-18 "
        data-icon="exclamation-triangle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          fill="currentColor"
        />
      </svg>
      <span
        class="textStatus"
      >
        Active
        <span
          class="versionNumber"
        >
          No longer available
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
